### ISSUE TITLE
Default function parameters

### ISSUE THREAD

**Issue by you-win** [godot-rust/gdext#381]
Allow for exposing default function parameters to GDScript.

The proposed syntax would look like:
```rs
#[godot_api]
impl MyType {
    #[func(defaults = [ 1 ]]
    fn some_func(a: Option<i32>) -> i32 {
        a.unwrap_or(1)
    }
}
```

This would map roughly to this in C++:
```c++
ClassDB::bind_method(D_METHOD("some_func", "a"), &MyType::some_func, DEFVAL(1));
```

Ideally the macro would also allow for function calls from Godot to automatically insert the default parameters as well so that `Option<T>` is not required, but I'm not sure if that's feasible.

WIP in #380

**Comment by Bromeon** [comment]
Thanks! I'll answer here instead of the PR, to first discuss the design before delving into implementation.

---

## 1. Function vs. parameter attributes

I wonder if an attribute on the parameter itself is not more intuitive. There are multiple options to do it syntactically.
```rs
    #[func]
    fn some_func(
        #[opt(default = 20)] a: i32
    ) -> i32 {
        ...
    }
```
Here the advantage is not significant, but for long parameter lists and more complex expressions like `Vector3i(1, 2, 3)`, it can make a big difference.

---

## 2. `Option<T>` vs `T`

> Ideally the macro would also allow for function calls from Godot to automatically insert the default parameters as well so that `Option<T>` is not required, but I'm not sure if that's feasible.

I think that should be possible. Maybe both can be allowed:
```rs
    #[func]
    fn some_func(
        #[opt] a: Option<i32>,
        #[opt = "str"] s: GodotString,
    ) -> i32 {
        let a = a.unwrap_or(1);
    }
```

---

## Magic needed?

First I was thinking to always infer `Option` as "optional", but that's probably not a good idea because 1) it's not explicit, 2) for some types (objects) it represents `nil`, not "optional".

But then, do we really need default values to be encoded inside an attribute?
Instead of:
```rs
    #[func]
    fn some_func(
        #[opt = 20] a: i32,
        #[opt = 15] s: i32,
    ) -> i32 {
        ...
    }
```
why not this?
```rs
    #[func]
    fn some_func(
        #[opt] a: Option<i32>,
        #[opt] b: Option<i32>,
    ) -> i32 {
        let a = unwrap_or(20);
        let b = unwrap_or(15);
    }
```
While a bit more verbose, it's immediately clear what happens and also allows for more complex relations, for example:
```rs
    #[func]
    fn some_func(
        #[opt] a: Option<i32>,
        #[opt] b: Option<i32>,
    ) -> i32 {
        let a = a.or(b).unwrap_or(44);
    }
```

**Comment by you-win** [comment]
Regarding your points:

### 1. Function vs. parameter attributes

I agree, this looks better syntactically. It does require additional verification during parsing, since I believe Godot displays default values like:

C++ binding
```cpp
ClassDB::bind_method(D_METHOD("some_func", "a", "b"), &MyType::some_func, DEFVAL(1));
```

GDScript Doc
```gdscript
func some_func(a, b = 1) int
```

So if the attribute were implemented per parameter, it would also be necessary to check that optionals are consecutive and that all required parameters come first. It's definitely doable and more ergonomic but also increases parsing complexity.

### 2. Option<T> vs T

Spliting this into 2 subpoints:

#### 2.1 Inserting the default values when they are not provided by Godot

I like the `#[opt = "str"]` syntax, however I'm trying to envision how this data would be registered in [`varcall`](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L116). I think it's definitely possible like with how parameters are automatically placed into a tuple, but I'm not seeing it quite yet.

#### 2.2. Magically inferring the defaults

I think this would be great, but at some point, doesn't this mean the proc macro would have to contain its own mini Rust parser? This is because the method bind function requires explicit defaults to be passed.

Given the cases below:

```rs
fn some_func(&self, #[opt] first: Option<i32>, #[opt] second: Option<i32>) {
    // Probably easy to extract the default
    // Note that this is the second parameter
    let second = second.unwrap_or(44);

    // How to automatically extract the default here?
    // The method bind will be incorrect if the default cannot be determined, so throw an error instead?
    other_related_func_that_takes_option(first);

    // This would also be difficult to determine
    let first = first.unwrap_or_else(|| self.stored_fn_call());
    // Different ways of providing a default
    let first = match first {
        Some(v) => v,
        None => 2,
    }
    let first = if let Some(v) = first {
        v
    } else {
        3
    };
    let first = if second.is_some() {
        second.unwrap();
    } else {
        4
    };
}
```

Magically determining the default could be restricted to only certain ways of handling options (e.g. the option must be handled using `unwrap_or` at the top of the fn body), but that doesn't feel great and would be potentially confusing if one didn't realize that's how default params are handled.

I think it would be magical and incredibly ergonomic if defaults could be determined automatically, but I think it would also be more trouble than it's worth.

**Comment by lilizoey** [comment]
i feel like `Option<T>` is a bit weird to symbolize a default argument, shouldn't this just mean that you can pass in either nil or a value of type `T`? so shouldn't an argument of type `i32` with a default value still just be an `i32`. since you'll always get some value of type `i32`? the main difference here would be that you can't tell whether you were passed a value explicitly or not, but like you can't tell that in gdscript either.

i feel like default values should just use the `Default` impl with an option to override it. We could maybe also make a type like
```rs
struct DefaultI32<const N: i32> {
  value: i32
}

impl<const N: i32> Default for DefaultI32<N> {
  fn default() -> Self {
    Self { value: N }
  }
}
```
for some cases.

**Comment by lilizoey** [comment]
> why not this?
> 
> ```rust
>     #[func]
>     fn some_func(
>         #[opt] a: Option<i32>,
>         #[opt] b: Option<i32>,
>     ) -> i32 {
>         let a = unwrap_or(20);
>         let b = unwrap_or(15);
>     }
> ```
> 
> While a bit more verbose, it's immediately clear what happens and also allows for more complex relations, for example:
> 
> ```rust
>     #[func]
>     fn some_func(
>         #[opt] a: Option<i32>,
>         #[opt] b: Option<i32>,
>     ) -> i32 {
>         let a = a.or(b).unwrap_or(44);
>     }
> ```

this goes against rust philosophy. in rust, the signature of a function generally does not depend on the body of a function (yes there is `impl Trait` but that's basically the one exception). and here you're suggesting we change the signature of a function we expose to godot to depend on the body of the function. i do not like that. neither gdscript nor rust does this, i dont think we should either.

**Comment by you-win** [comment]
> i feel like Option<T> is a bit weird to symbolize a default argument, shouldn't this just mean that you can pass in either nil or a value of type T?

The difficulty with _not_ using `Option<T>` is that it's impossible to tell if a developer-provided default should be used. i.e. i32::default() is just 0, but 0 is also a valid input.

The wrapper type for `Default` makes sense, but I wonder if there's a better way.

> the main difference here would be that you can't tell whether you were passed a value explicitly or not, but like you can't tell that in gdscript either

This is possible in Rust (and C++ with vararg methods) since Godot passes along an [`arg_count` parameter for varcalls](https://github.com/godot-rust/gdext/blob/master/godot-macros/src/method_registration/register_method.rs#L100). With how varcalls are implemented in gdext, the [expected parameter count](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L12) is also available when [converting params from Godot -> params for Rust](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L166).

My WIP PR modifies several signatures to pass along the `arg_count`, but what to do afterwards is still an open question. Whether it be a wrapper type for `Default`, [inserting default values into the varcall arg converter somehow](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L149), requiring default params to be `Option<T>`, or something else.

**Comment by lilizoey** [comment]
> > i feel like Option is a bit weird to symbolize a default argument, shouldn't this just mean that you can pass in either nil or a value of type T?
> 
> The difficulty with _not_ using `Option<T>` is that it's impossible to tell if a developer-provided default should be used. i.e. i32::default() is just 0, but 0 is also a valid input.

I'm not sure why that's needed?

> 
> > the main difference here would be that you can't tell whether you were passed a value explicitly or not, but like you can't tell that in gdscript either
> 
> This is possible in Rust (and C++ with vararg methods) since Godot passes along an [`arg_count` parameter for varcalls](https://github.com/godot-rust/gdext/blob/master/godot-macros/src/method_registration/register_method.rs#L100). With how varcalls are implemented in gdext, the [expected parameter count](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L12) is also available when [converting params from Godot -> params for Rust](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L166).
> 
> My WIP PR modifies several signatures to pass along the `arg_count`, but what to do afterwards is still an open question. Whether it be a wrapper type for `Default`, [inserting default values into the varcall arg converter somehow](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L149), requiring default params to be `Option<T>`, or something else.

Yeah it is technically possible in rust, but why would we want people to be able to treat the case of "no argument passed so use the default value" and "default value explicitly passed to the function" as different? to me that honestly seems rather weird and unintuitive. Like i'd expect that those two cases behave identically anyway.

**Comment by you-win** [comment]
>>> i feel like Option is a bit weird to symbolize a default argument, shouldn't this just mean that you can pass in either nil or a value of type T?
>>
>> The difficulty with not using Option<T> is that it's impossible to tell if a developer-provided default should be used. i.e. i32::default() is just 0, but 0 is also a valid input.
>
> I'm not sure why that's needed?

It's needed since the dev could define a different default from what Rust considers `Default`.

>>> the main difference here would be that you can't tell whether you were passed a value explicitly or not, but like you can't tell that in gdscript either
>>
>> This is possible in Rust (and C++ with vararg methods) since Godot passes along an [arg_count parameter for varcalls](https://github.com/godot-rust/gdext/blob/master/godot-macros/src/method_registration/register_method.rs#L100). With how varcalls are implemented in gdext, the [expected parameter count](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L12) is also available when [converting params from Godot -> params for Rust](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L166).
>>
>> My WIP PR modifies several signatures to pass along the arg_count, but what to do afterwards is still an open question. Whether it be a wrapper type for Default, [inserting default values into the varcall arg converter somehow](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L149), requiring default params to be Option<T>, or something else.
>
> Yeah it is technically possible in rust, but why would we want people to be able to treat the case of "no argument passed so use the default value" and "default value explicitly passed to the function" as different? to me that honestly seems rather weird and unintuitive. Like i'd expect that those two cases behave identically anyway.

Rust doesn't support default params, so there needs to be a way to determine if a default should be used. If the standard `Default` impl was used, there would be no way to tell in the Rust `fn` if this is developer input or `Default::default()`.

That's why I brought up `Option<T>` as a potential solution, which means the default value would be defined in the actual `fn` body. Other solutions, like a Default wrapper type, would insert the developer's default value during the varcall invocation. Although thinking about it more, the wrapper type wouldn't be able to support defaults per function, which is something that Godot supports.

For example:

Using Default (the current implementation):
```rs
unsafe fn varcall_arg<P: FromVariant + Default, const N: isize>(
    args_ptr: *const sys::GDExtensionConstVariantPtr,
    arg_count: isize,
    method_name: &str,
) -> P {
    if arg_count >= N {
        return P::default();
    }
    let variant = &*(*args_ptr.offset(N) as *mut Variant); // TODO from_var_sys
    P::try_from_variant(variant)
        .unwrap_or_else(|_| param_error::<P>(method_name, N as i32, variant))
}

#[func(defaults = [4]]
fn foo(a: i32) {
    ... do stuff ...
}

#[func(defaults = [10]]
fn bar(a: i32) {
    ... how to handle this case? ...
}
```

vs

Using Option (assuming `None` is passed if a parameter is missing, using the current macro implementation):
```rs
unsafe fn varcall_arg<P: FromVariant, const N: isize>(
    args_ptr: *const sys::GDExtensionConstVariantPtr,
    arg_count: isize,
    method_name: &str,
) -> P {
    if arg_count >= N {
        return None;
    }
    let variant = &*(*args_ptr.offset(N) as *mut Variant); // TODO from_var_sys
    P::try_from_variant(variant)
        .unwrap_or_else(|_| param_error::<P>(method_name, N as i32, variant))
}

#[func(defaults = [4]]
fn foo(a: Option<i32>) {
    let a = a.unwrap_or(4);
}

#[func(defaults = [10]]
fn bar(a: Option<i32>) {
    let a = a.unwrap_or(10);
}
```

The point is, it's possible to tell where default values should be used. The question is, where should they be handled. I should also mention that Godot does _not_ send any default values, it just sends less parameters.

**Comment by Bromeon** [comment]
> i feel like default values should just use the `Default` impl with an option to override it. We could maybe also make a type like
> 
> ```rust
> struct DefaultI32<const N: i32> {
>   value: i32
> }
> 
> impl<const N: i32> Default for DefaultI32<N> {
>   fn default() -> Self {
>     Self { value: N }
>   }
> }
> ```

The Rust `Default` trait keeps getting mixed up with anything that has "default" in the name, but it really doesn't fit here. 

Traits have one possible implementation per type. What we want is one implementation per parameter.

I don't think hacking this together via const generics would be the way to go 😉 

---

> [you-win] I think this would be great, but at some point, doesn't this mean the proc macro would have to contain its own mini Rust parser? This is because the method bind function requires explicit defaults to be passed.

> [lilizoey] this goes against rust philosophy. in rust, the signature of a function generally does not depend on the body of a function (yes there is `impl Trait` but that's basically the one exception). and here you're suggesting we change the signature of a function we expose to godot to depend on the body of the function. i do not like that. neither gdscript nor rust does this, i dont think we should either.

You both misunderstood that or I didn't express myself clearly; we would not analyze the body to infer any default arguments. 

The idea was, since Rust can detect whether a value was provided or not, we could use this to pass `Some(val)` or `None` to a user function accepting `Option<T>`. This could be covered by a simple `unwrap_or()` and would avoid more proc-macro DSL. No value appearing in the body is advertised.

But there are some caveats, you mentioned some of them already:

* We cannot display a default value in GDScript docs (or even choose a meaningful one).
* It's not obvious for the caller that passing nothing is different from passing the default value. In other words, the system works a bit different from other GDScript functions with default parameters, which may be confusing.
* There is potential confusion with `null`.

So it's probably better to build something with `#[opt]`. However I'd like to keep complexity low -- too much magic in proc-macros has a lot of downsides.

**Comment by you-win** [comment]
> The idea was, since Rust can detect whether a value was provided or not, we could use this to pass Some(val) or None to a user function accepting Option<T>.

How would `val` be defined if the only macro is `#[opt]`? `val` needs to be defined somewhere.

> We cannot display a default value in GDScript docs (or even choose a meaningful one).

We can, I mentioned that it's possible in an [earlier comment](https://github.com/godot-rust/gdext/issues/381#issuecomment-1676359576). The only issue is that they are displayed in reverse list order, if that makes sense.

**Comment by Bromeon** [comment]
Let's say we forget about `Option`, `Default`, wrapper types or anything else.

Based on my previous suggestion, what about this?
```rs
#[func]
fn some_func(
    &self,
    required: i32,
    #[opt = -20] optional: i32,
    #[opt = "hello"] optional2: GodotString,
) {
    // All parameters have defined values here, in all cases.
}
```
Could be called from GDScript as:
```gdscript
obj.some_func(1)
obj.some_func(1, 2)
obj.some_func(1, 2, "hi")
```

**Comment by Bromeon** [comment]
> How would `val` be defined if the only macro is `#[opt]`? `val` needs to be defined somewhere.

That's a good point, but also there's a point in @lilizoey's argument that it should be possible for a GDScript caller to pass an argument that has the same semantics as the default value. And there may be cases where `Option<T>` parameters can be assigned `null` arguments.

So I'm not sure if we'd even want `#[opt]` stand-alone.




> We can, I mentioned that it's possible in an [earlier comment](https://github.com/godot-rust/gdext/issues/381#issuecomment-1676359576). The only issue is that they are displayed in reverse list order, if that makes sense.

I meant that we can only do that if the default value is part of the signature (including attribute). If it's in the body as `unwrap_or`, we can't.

**Comment by lilizoey** [comment]
> Rust doesn't support default params, so there needs to be a way to determine if a default should be used. If the standard `Default` impl was used, there would be no way to tell in the Rust `fn` if this is developer input or `Default::default()`.
> 
> That's why I brought up `Option<T>` as a potential solution, which means the default value would be defined in the actual `fn` body. Other solutions, like a Default wrapper type, would insert the developer's default value during the varcall invocation. Although thinking about it more, the wrapper type wouldn't be able to support defaults per function, which is something that Godot supports.

I think we're talking past each other a bit maybe. Yes, we'll likely need to use `Option<T>` somewhere, for instance to communicate whether a default argument should be used or not. What im arguing is that the user of the function shouldn't define their function to take an `Option<T>` to signal this. (of course unless they actually intend `None` to be a valid value for this argument)

We would simply have some code in-between the user-defined function and the godot call. But the *user* of the function would not see any of this. To them a default argument is just that, a default for when the argument is left out. I dont remember exactly where this would be in the code gen, but we already dont just pass the function pointers directly to godot to be called, we can code-gen the default argument value in between there.

**Comment by lilizoey** [comment]
> Let's say we forget about `Option`, `Default`, wrapper types or anything else.
> 
> Based on my previous suggestion, what about this?
> 
> ```rust
> #[func]
> fn some_func(
>     &self,
>     required: i32,
>     #[opt = -20] optional: i32,
>     #[opt = "hello"] optional2: GodotString,
> ) {
>     // All parameters have defined values here, in all cases.
> }
> ```
> 
> Could be called from GDScript as:
> 
> ```gdscript
> obj.some_func(1)
> obj.some_func(1, 2)
> obj.some_func(1, 2, "hi")
> ```
> 
> We'd need to keep in mind that if we do it via proc-macros, there's a limit to how complex the expressions can get, at least if they should stay debuggable. But it's probably an OK trade-off.

should `#[opt] optional: i32` be allowed? if so then i feel it'd make sense for that to be equivalent to `#[opt = i32::default()] optional: i32`

**Comment by Bromeon** [comment]
> should `#[opt] optional: i32` be allowed? if so then i feel it'd make sense for that to be equivalent to `#[opt = i32::default()] optional: i32`

I would probably start without it, as it tends to weaken type safety (people abuse empty strings etc. to denote absence of a value). 

Syntax is also up to bikeshedding 😀 for fields, we already have `#[init(default = ...)]`, and generally, our `KvParser` doesn't allow top-level key-value pairs like `#[opt = value]`. Maybe we could make an exception for parameters, but I'd also like to explore some alternatives 🤔

**Comment by you-win** [comment]
>That's a good point, but also there's a point in lilizoey's argument that it should be possible for a GDScript caller to pass an argument that has the same semantics as the default value. And there may be cases where Option<T> parameters can be assigned null arguments.

Okay that's becoming more clear. There are instances where GDScript `null` could be passed as a valid parameter.

>I dont remember exactly where this would be in the code gen, but we already dont just pass the function pointers directly to godot to be called, we can code-gen the default argument value in between there.

As mentioned [previously](https://github.com/godot-rust/gdext/issues/381#issuecomment-1676439876), the [codegen is here](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L116) for parameters, so inserting user-defined defaults here might make sense.

>should #[opt] optional: i32 be allowed? if so then i feel it'd make sense for that to be equivalent to #[opt = i32::default()] optional: i32

If that's meant to be shorthand for "use whatever the Rust default is", I'm for it. I just don't want this case to be forgotten (using new possible syntax):

```rs
fn foo(
    #[opt(1)] a: i32, // defaults to 1
    #[opt(2)] b: i32, // defaults to 2
    #[opt] c: i32 // defaults to 0
) {}
```

**Comment by lilizoey** [comment]
> > That's a good point, but also there's a point in lilizoey's argument that it should be possible for a GDScript caller to pass an argument that has the same semantics as the default value. And there may be cases where Option parameters can be assigned null arguments.
> 
> Okay that's becoming more clear. There are instances where GDScript `null` could be passed as a valid parameter.
> 
> > I dont remember exactly where this would be in the code gen, but we already dont just pass the function pointers directly to godot to be called, we can code-gen the default argument value in between there.
> 
> As mentioned [previously](https://github.com/godot-rust/gdext/issues/381#issuecomment-1676439876), the [codegen is here](https://github.com/godot-rust/gdext/blob/master/godot-core/src/builtin/meta/signature.rs#L116) for parameters, so inserting user-defined defaults here might make sense.

The codegen i was talking about is in fact [here](https://github.com/godot-rust/gdext/blob/b4e6fd6c9c23e0fd3a124c2893a9592c861ab780/godot-macros/src/method_registration/mod.rs#L53). since this generates something like:
```rs
|_, params| {
  let (a,) = params;
  foo(a)
}
```
and in here we can just have this become something like

```rs
|_, params, default_params| {
  let (a,) = params;
  let (b,c,...) = default_params;
  foo(a, b.unwrap_or(b_default), c.unwrap_or(c_default), ...)
}
```

where the `b_default` and `c_default` and so on are the values defined in `#[opt = b_default]`, and the values in the `default_params` tuple are all `Option<T>` for each parameter.

**Comment by you-win** [comment]
That makes sense. So we can split the `params` tuple by required/optional, and then in varcall, implicitly wrap every default param in `Option`?

**Comment by lilizoey** [comment]
That was my idea at least, i definitely do think we should treat required and optional arguments separately. Since i feel like it'd be very messy to try to unify them into a single list of arguments somehow.

**Comment by Bromeon** [comment]
> What I like about `Option<T>` to signal that your argument is optional and a default value will be provided, is that it's rather intuitive.

A subtle difference exists though: `Option` currently means "nullable", not "possibly absent".

There are definitely cases where people want it explicit:
```gdscript
    ship.set_target(null)
    # and not:
    ship.set_target()
```

Also, `Option<T>` can appear anywhere in a `#[func]` parameter list. It would be strange if it can be omitted when at the end, but not somewhere in between.

Something along `#[opt]` looks nice. The `#[opt = defaultValue]` was suggested, but the only advantage it has is that the default value appears in GDScript docs? I'm not sure if that's worth the complexity, or are there more benefits? Otherwise we can start simple, let the user use `unwrap_or()` and only extend the API on demand.

**Comment by astrale-sharp** [comment]
> A subtle difference exists though: `Option` currently means "nullable", not "possibly absent".

I didn't think about that, I'm convinced.

```rust
#[func] fn set_integer(
    #[opt(default = defaultValue)] i: i32
);

#[func] fn set_integer(
    #[opt] i: Option<i32>
);
```

I like these two options best, let's definitely start simple though, I know how it is with proc macros now and I feel like small PRs are more easily made with quality~~

**Comment by astrale-sharp** [comment]
implementation wise, should I try to have a modified `varcall_arg` function named `varcall_arg_default -> Option<R>`  so that if there is nothing at `*args_ptr.offset(N)` (which is to say that args_len < N) it then returns None?

Then  `PtrcallSignatureTuple::in_ptrcall`, if `#[opt]` was present uses that function?

**Comment by Zitronenjoghurt** [comment]
Any news on this feature?

**Comment by Bromeon** [comment]
No

**Comment by astrale-sharp** [comment]
A plan

`ClassMethodInfo::from_signature` is only used once in the code base so we could add an argument to this method (uncomment //default argument ;) )
it's a Vec<Variant>, I'm not sure if ensuring that we pick a variant of the right type is enough but nothing is currently mentioned about it in the current safety comment.
Then in
`godot-macros/src/class/data_models/inherent_impl.rs:80:` `transform_inherent_impl`
in the call of 
`godot-macro/src/class/data_models/func.rs:92 make_method_registration`
add a new function called extract_default_arguments taking func_definition that 
- returns `ParseResult<TokenStream>`
- Ensures all the opt attributes are applied to the last arguments (or fail)
- returns a token stream of the form `quote! { vec! [godot::builtin::Variant::from(#extracted_default)] }`

Use that stream to feed the `ClassMethodInfo::from_signature` call


LEFT TO BE DONE, update the documentation accordingly if godot doesn't do it
Note: 
`godot-cpp 61f52` `gdextension/gdextension_interface.h:466:3 ` `GDExtensionClassMethodInfo;defaut_argument_count`
is not really documented so I assume it means only the last n arguments can have a default.

I'd be keen to deal with this.
Good day!

**Comment by Bromeon** [comment]
@astrale-sharp sounds good! I'd probably need to see the code directly to see how things play together 🙂

We should also agree on the user-facing API before jumping to implementation.
I'd suggest to start with an MVP (minimal viable product):
```rs
#[func] 
fn some_func(
    required: i32,
    #[opt(default = "hello")] string: GString,
    #[opt(default = 123)] integer: i32,
    #[opt(default = None)] object: Option<Gd<Node>>
);
```

The `default` key would be required for now, and the value specified would have to implement `AsArg<ParameterType>`. This allows `"hello"`  to be used for `GString`, instead of `GString::from("hello")`.

In a possible later step -- depending on need, I'd explicitly exclude this from the first PR -- we could then see if we want to support `Option`, with only the `#[opt]` attribute. This would enable users to explicitly check whether an argument was provided or not.
```rs
#[func] 
fn some_func(
    required: i32,
    #[opt] string: Option<GString>,
    #[opt] integer: Option<i32>,
    #[opt] object: Option<Gd<Node>>
);
``` 
The `Option<Gd>` is an interesting case, as the user can technically not differentiate between nothing being passed, or `null`  being passed. It would need `Option<Option<Gd<Node>>` for that. But I don't think any of this matters in practice.

**Comment by astrale-sharp** [comment]
I'm quite ok with the user facing API in your suggestion which is supported directly in KvParser.

It feels right to `modify into_signature_info` and `SignatureInfo` to store the parameter parsing information as I'd argue it doesn't belong in `godot-macros/src/class/data_models/inherent_impls.rs:517` which seems to be related to func or items only attributes.

It does mean that into_signature_info will become faillible because we would parse in it.

Unfortunately signature_info is used in `GetterSetterImpl` and `handle_regular_virtual_function` in `godot-macros` which we might not want to make faillible.

We could alternatively use a different type than `FunctionDef` which hold `FuncDefAlt` with an alternative `SignatureInfoAlt` with more info and faillible on parsed. ofc the names are not to stay and i don't find this option elegant at all.

I've worked a bit on this to explore the various problems that might arise.

I'm thinking of using `<default_parameter as AsArg>.to_variant()` to feed `ClassMethodInfo::from_signature`

**Comment by ttencate** [comment]
Fwiw, I'm a fan of the syntax chosen so far.

> In a possible later step -- depending on need, I'd explicitly exclude this from the first PR -- we could then see if we want to support `Option`, with only the `#[opt]` attribute. This would enable users to explicitly check whether an argument was provided or not.

Would this use `Default::default()`, or have a special case for only `Option::None`? I think using `Default` would be a sensible extension with no downsides. As a user that's definitely what I'd expect to happen.

Open question is how this interacts with `AsArg` but I'm sure that can be figured out.

**Comment by Bromeon** [comment]
@astrale-sharp it's a bit hard to follow from files and line numbers. If you are already experimenting, feel free to open a draft PR and we can iterate on it 🙂 

@ttencate `Default` is possible, although several types don't support it. One possibility would be to have it similar to `#[init(val = ...)]` in fields. I was even thinking about `#[opt(val = ...)]` but I'm not sure if it's clear enough? It's different from fields, which are always initialized with that value, while for parameters it's only a fallback.

Either way, I'd start with a minimal API and then extend it depending on needs 😉 


> Open question is how this interacts with `AsArg` but I'm sure that can be figured out.

Ironically, integrating with `AsArg` usually makes a problem 100 times harder 🙈 see #1193 for example. But I think it would only be needed for the values provided via `#[opt(default = ...)]`, whereas `#[opt]` itself could work with types directly.

**Comment by ttencate** [comment]
> `Default` is possible, although several types don't support it.

Which ones?

**Comment by ttencate** [comment]
> But I think it would only be needed for the values provided via `#[opt(default = ...)]`, whereas `#[opt]` itself could work with types directly.

Yeah, now that I've woken up fully, I see that there is no other way. `#[opt] foo: GString` would require `GString` to implement `Default`, and couldn't go through `AsArg` at all because there is no other type to convert _from_.

**Comment by Bromeon** [comment]
> Which ones?

Everything that has no meaningful default value: `Gd` for non-default-constructible or manually allocated classes, enums, `Plane`, `Signal`, `Callable` and `Rid`. See also [rationale](https://godot-rust.github.io/docs/gdext/master/godot/__docs/index.html#3-default-trait-only-when-the-default-value-is-common-and-useful) 🙂

**Comment by ttencate** [comment]
Agreed that for each of those it makes sense to be explicit when specifying default parameters, especially when manual allocations are involved. So by that argument too, using `Default` for a bare `#[opt]` would be a... useful default.

**Comment by astrale-sharp** [comment]
> [@astrale-sharp](https://github.com/astrale-sharp) it's a bit hard to follow from files and line numbers. If you are already experimenting, feel free to open a draft PR and we can iterate on it 🙂

Done! Please don't consider it anything close to ready to, I'm aware most of it doesn't work, but hopefully it makes my comment about `FuncDef` and `SignatureInfo` clearer

**Comment by astrale-sharp** [comment]
Okay I've been thinking about this for a bit and i guess the solution that makes the most sense is for signature info to contain parameters attributes (not faillible) and the verification/parsing etc can just be done later.

**Comment by Bromeon** [comment]
Let's discuss implementation details on the PR please, and keep this issue for general design considerations.
Too much context switching otherwise 🙂
